uaclient/lock.py

```diff
@@ -0,0 +1,89 @@
+import functools
+import logging
+import os
+import time
+
+from uaclient import config, exceptions
+
+LOG = logging.getLogger("ua.lock")
+
+# Set a module-level callable here so we don't have to reinstantiate
+# UAConfig in order to determine dynamic data_path exception handling of
+# main_error_handler
+clear_lock_file = None
+
+
+def clear_lock_file_if_present():
+    global clear_lock_file
+    if clear_lock_file:
+        clear_lock_file()
+
+
+class SingleAttemptLock:
+    """
+    Context manager for gaining exclusive access to the lock file.
+    Create a lock file if absent. The lock file will contain a pid of the
+        running process, and a customer-visible description of the lock holder.
+    :param lock_holder: String with the service name or command which is
+        holding the lock.
+        This lock_holder string will be customer visible in status.json.
+    :raises: LockHeldError if lock is held.
+    """
+
+    def __init__(self, cfg: config.UAConfig, lock_holder: str):
+        self.cfg = cfg
+        self.lock_holder = lock_holder
+
+    def __enter__(self):
+        global clear_lock_file
+        (lock_pid, cur_lock_holder) = self.cfg.check_lock_info()
+        if lock_pid > 0:
+            raise exceptions.LockHeldError(
+                lock_request=self.lock_holder,
+                lock_holder=cur_lock_holder,
+                pid=lock_pid,
+            )
+        self.cfg.write_cache(
+            "lock", "{}:{}".format(os.getpid(), self.lock_holder)
+        )
+        notice_msg = "Operation in progress: {}".format(self.lock_holder)
+        self.cfg.add_notice("", notice_msg)
+        clear_lock_file = functools.partial(self.cfg.delete_cache_key, "lock")
+
+    def __exit__(self, _exc_type, _exc_value, _traceback):
+        global clear_lock_file
+        self.cfg.delete_cache_key("lock")
+        clear_lock_file = None  # Unset due to successful lock delete
+
+
+class SpinLock(SingleAttemptLock):
+    def __init__(
+        self,
+        cfg: config.UAConfig,
+        lock_holder: str,
+        sleep_time: int = 10,
+        max_retries: int = 12,
+    ):
+        super().__init__(cfg, lock_holder)
+        self.sleep_time = sleep_time
+        self.max_retries = max_retries
+
+    def __enter__(self):
+        LOG.debug("spin lock starting for {}".format(self.lock_holder))
+        tries = 0
+        done = False
+        while not done:
+            try:
+                super().__enter__()
+                done = True
```

Comment:
I think we can just use a break here instead

Reply:
good point!